client/src/main/java/io/pravega/client/stream/impl/EventStreamReaderImpl.java

```diff
@@ -292,7 +295,8 @@ private void handleSegmentTruncated(EventSegmentReader segmentReader) throws Tru
         String delegationToken = groupState.getOrRefreshDelegationTokenFor(segmentId);
 
         @Cleanup
-        SegmentMetadataClient metadataClient = metadataClientFactory.createSegmentMetadataClient(segmentId, delegationToken);
+        SegmentMetadataClient metadataClient = metadataClientFactory.createSegmentMetadataClient(segmentId,
+                DelegationTokenProviderFactory.create(delegationToken, controller, segmentId));
```

Comment:
This is sub optimal. If this provider refreshes the token the delegationToken in this class will still be out of date, so every future call to this method is going to have to refresh the token again even though it just retrieved a token that has not expired.

Reply:
Sure, I've made the change that you've suggested in the latest commit.